aws-s3-transfer-manager/src/operation/upload_objects.rs

```diff
@@ -24,14 +32,56 @@ pub(crate) struct UploadObjects;
 impl UploadObjects {
     /// Execute a single `UploadObjects` transfer operation
     pub(crate) async fn orchestrate(
-        _handle: Arc<crate::client::Handle>,
-        _input: UploadObjectsInput,
-    ) -> Result<UploadObjectsHandle, UploadObjectsError> {
-        unimplemented!()
+        handle: Arc<crate::client::Handle>,
+        input: UploadObjectsInput,
+    ) -> Result<UploadObjectsHandle, crate::error::Error> {
+        //  validate existence of source and return error if it's not a directory
+        let source = input.source().expect("source set");
+        validate_target_is_dir(source).await?;
+
+        let concurrency = handle.num_workers();
+        let ctx = UploadObjectsContext::new(handle.clone(), input);
+
+        // spawn all work into the same JoinSet such that when the set is dropped all tasks are cancelled.
+        let mut tasks = JoinSet::new();
+        let (work_tx, work_rx) = async_channel::bounded(concurrency);
+
+        // spawn worker to discover/distribute work
+        tasks.spawn(worker::list_directory_contents(
+            ctx.state.input.clone(),
+            work_tx,
+        ));
+
+        for i in 0..concurrency {
+            let worker = worker::upload_objects(ctx.clone(), work_rx.clone())
+                .instrument(tracing::debug_span!("object-uploader", worker = i));
+            tasks.spawn(worker);
+        }
+
+        let handle = UploadObjectsHandle { tasks, ctx };
+        Ok(handle)
     }
 }
 
-/// Error type for `UploadObjects` operation
-#[non_exhaustive]
+/// DownloadObjects operation specific state
 #[derive(Debug)]
-pub enum UploadObjectsError {}
+pub(crate) struct UploadObjectsState {
+    input: UploadObjectsInput,
+    failed_uploads: Mutex<Option<Vec<FailedUploadTransfer>>>,
```

Comment:
I think we have discussed this before, but why wrap this in an `Option`? What will be the difference between an empty vector and `None`? Can we just get rid of Option here?

Reply:
The change itself was to address [TODO](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/main/aws-s3-transfer-manager/src/operation/upload_objects/output.rs#L17) here, but I see the point you're trying to make. For a struct, like a builder, where a user configures field values, it can be useful to have `Option<Vec<...>>` where `None` indicates the user didn't supply a custom value, and `Some(vec![])` indicates the user explicitly passed-in an empty vec. But in this particular case, probably there isn't a difference between `Option<Vec<...>>` vs. `Vec<...>`, as the user probably won't construct `UploadObjectsOutput`.

I'm ok with getting rid of `Option` here, and we need to do so consistently for [DownloadObjectsState](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/790ead476a104cf0b66fdd00b5b9c3636321b244/aws-s3-transfer-manager/src/operation/download_objects.rs#L95) as well. I'd like to hear what others think before proceeding with it.

Reply:
Removed `Option` from both download and upload in de6d920.